python/pyarrow/tests/test_parquet.py

```diff
@@ -1958,8 +2069,11 @@ def _partition_test_for_filesystem(fs, base_path):
     expected_df = (df.sort_values(by='index')
                    .reset_index(drop=True)
                    .reindex(columns=result_df.columns))
-    expected_df['foo'] = pd.Categorical(df['foo'], categories=foo_keys)
-    expected_df['bar'] = pd.Categorical(df['bar'], categories=bar_keys)
+    if not use_dataset:
+        # TODO(dataset) Dataset API does not create categorical columns
+        # for partition keys
```

Comment:
Is this critical? This would be fairly easy to recover with an option like `discover_dictionaries` in partition schema discovery

Reply:
Similar as above: I would prefer that we first decice what we want long term, rather than exactly mimicking the old API (certainly given it is opt-in for now).

Reply:
Given the variable dictionary support we have now, returning dictionary encoded fields shouldn't (in principle?) be too hard and also should be faster than duplicating a string many times. In the worst case, a record batch would have a field having all 0 indices and a dictionary with a single value.